### PR TITLE
Fix: Add Stoplight to guard erronous Web::PushSubscriptions

### LINKED ISSUE #32504: Guard WebPushSubscriptions with Stoplight
### Pitch

Currently we have no handling for persistent errors with WebPush subscriptions. We should likely add stoplight code in the WebPush delivery worker like what we have in the ActivityPub delivery worker.

### Motivation

This would reduce the failure rates in redis by simply discarding the jobs when the push subscription is in an unhealthy state.

### PR DESCRIPTION
**Pull request by ThisIsMissEm**
This fixes #32504, I've based it on the code for DeliveryWorker

The idea being that if a push subscription is failing continuously, we should back off and not keep retrying it for subsequent notifications.

It's worth noting this won't prevent sidekiq failures, but will drastically reduce the number of them for a badly behaving endpoint.

### CODE REVIEW

File: app/workers/web/push_notification_worker.rb [hunk 1]
```diff
@@ -8,6 +8,9 @@ class Web::PushNotificationWorker
   TTL     = 48.hours.to_s
   URGENCY = 'normal'
 
+  STOPLIGHT_FAILURE_THRESHOLD = 10
+  STOPLIGHT_COOLDOWN = 60
```
Comment (ThisIsMissEm) [inline app/workers/web/push_notification_worker.rb:12]: I'm not sure if these are right for Web::PushSubscriptions, might need tuning

File: app/workers/web/push_notification_worker.rb [hunk 2]
```diff
@@ -18,9 +21,15 @@ def perform(subscription_id, notification_id)
 
     payload = web_push_request.encrypt(push_notification_json)
 
-    request_pool.with(web_push_request.audience) do |http_client|
-      request = Request.new(:post, web_push_request.endpoint, body: payload.fetch(:ciphertext), http_client: http_client)
+    perform_request(payload)
```
Comment (ThisIsMissEm) [inline app/workers/web/push_notification_worker.rb:24]: This refactoring is based on DeliveryWorker

File: app/workers/web/push_notification_worker.rb [hunk 3]
```diff
@@ -30,26 +39,27 @@ def perform(subscription_id, notification_id)
...
-        # and must be removed
-
-        if (400..499).cover?(response.code) && ![408, 429].include?(response.code)
-          @subscription.destroy!
-        elsif !(200...300).cover?(response.code)
-          raise Mastodon::UnexpectedResponseError, response
+  def perform_request(payload)
+    stoplight_wrapper.run do
+      request_pool.with(web_push_request.audience) do |http_client|
+        build_request(http_client, payload).perform do |response|
```
Comment (ThisIsMissEm) [inline app/workers/web/push_notification_worker.rb:48]: Is there anything we can do here for handling something like a firewall block, where the request is just dropped? This is a common failure mode for nfty

File: spec/workers/web/push_notification_worker_spec.rb [hunk 4]
```diff
@@ -37,17 +37,31 @@
...
     end
 
+    it 'raises when request fails' do
+      stub_request(:post, endpoint).to_return(status: 500)
+
+      expect { subject.perform(subscription.id, notification.id) }.to raise_error Mastodon::UnexpectedResponseError
+    end
+
+    it 'destroys the subscription if it returns 404' do
+      stub_request(:post, endpoint).to_return(status: 404)
```
Comment (ThisIsMissEm) [inline spec/workers/web/push_notification_worker_spec.rb:58]: Not just 404 triggers this behaviour, but also some other 4xx errors.

File: spec/workers/web/push_notification_worker_spec.rb [hunk 5]
```diff
@@ -37,17 +37,31 @@
...
 
       expect(web_push_endpoint_request)
         .to have_been_made
     end
 
+    it 'raises when request fails' do
+      stub_request(:post, endpoint).to_return(status: 500)
+
+      expect { subject.perform(subscription.id, notification.id) }.to raise_error Mastodon::UnexpectedResponseError
+    end
```
Comment (ThisIsMissEm) [inline spec/workers/web/push_notification_worker_spec.rb:55]: It feels like we should have some tests for the "stoplight did it's thing" but I don't see any example of those in the repository.

**Comment by ThisIsMissEm** [comment]
> > It's worth noting this won't prevent sidekiq failures, but will drastically reduce the number of them for a badly behaving endpoint.
> 
> It should not reduce the number of them, as Stoplight will raise an issue, and sidekiq will retry the job later as usual. However, the failure will happen immediately, without spending time querying a remote server.

Oh, okay. I didn't realise stoplight would re-raise the previous error, I thought if it was in the red state, then it'd just bypass that code.

In which case, I think that yes, we need something akin to DeliveryFailureTracker for both push subscriptions and webhooks.

**Comment by ThisIsMissEm** [comment]
@renchap / @ClearlyClaire let me know if you'd like to keep this PR open or close it, as I know you've an internal ticket for webpush errors now.

**Comment by renchap** [comment]
As discussed, closing this as this is not the correct fix. We are investigating adding support for per-subscription rate limits

**Pull request closed without merging**